resolwe/flow/models.py

```diff
@@ -583,32 +627,121 @@ def validation_schema(name):
+        if not data_qs.first()['process__type'].startswith(type_):
+            raise ValidationError(
+                "Referenced `Data` object is of wrong type (id:{})".format(data_pk))
+
+    for _schema, _fields, _ in iterate_schema(instance, schema):
+        name = _schema['name']
+
+        if test_required and _schema.get('required', False) and name not in _fields:
+            raise ValidationError("Required field \"{}\" not given.".format(name))
+
+        if name in _fields:
+            field = _fields[name]
+            type_ = _schema.get('type', "")
+
+            try:
+                jsonschema.validate([{"type": type_, "value": field}], TYPE_SCHEMA)
+            except jsonschema.exceptions.ValidationError as ex:
+                raise ValidationError(ex.message)
+
+            if type_ == 'basic:file:' and 'validate_regex' in _schema:
+                validate_file(field['file'], _schema['validate_regex'])
+
+            elif type_ == 'list:basic:file:' and 'validate_regex' in _schema:
+                for file_ in field:
+                    validate_file(file_['file'], _schema['validate_regex'])
+
+            elif type_ == 'basic:json:' and not Storage.objects.filter(pk=field).exists():
+                raise ValidationError(
+                    "Referenced `Storage` object does not exist (id:{})".format(field))
+
+            elif type_.startswith('data:'):
+                validate_data(field, type_)
+
+            elif type_.startswith('list:data:'):
+                for data_id in field:
+                    validate_data(data_id, type_)
+
+
 def _hydrate_values(output, output_schema, data):
     """Hydrate basic:file and basic:json values.
 
     Find fields with basic:file type and assign a full path to the file.
     Find fields with basic:json type and assign a JSON object from storage.
 
     """
+    def hydrate_file(file_name):
+        id_ = "{}/".format(data.id)  # needs trailing slash
+        if id_ in file_name:
+            file_name = file_name[file_name.find(id_) + len(id_):]  # remove id from filename
+
+        return os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], id_, file_name)
+
+    def hydrate_storage(storage_id):
+        if re.match('^[0-9a-fA-F]{24}$', str(storage_id)) is None:
+            print("ERROR: basic:json value in {} not ObjectId but {}.".format(name, storage_id))
+
+        return Storage.objects.get(pk=storage_id)
+
     for field_schema, fields in iterate_fields(output, output_schema):
         name = field_schema['name']
         value = fields[name]
         if 'type' in field_schema:
             if field_schema['type'].startswith('basic:file:'):
-                fn = value['file']
-                id_ = "{}/".format(data.id)  # needs trailing slash
-                if id_ in fn:
-                    fn = fn[fn.find(id) + len(id_):]  # remove id from filename
+                value['file'] = hydrate_file(value['file'])
```

Comment:
Add file size if there is no size key in the dict. Maybe write `hydrate_size()`?